File: app/modules/scoring/domain/end_state.py
"""
End-State Scoring - Portfolio-level scoring for holistic planning.

This module provides scoring functions that evaluate the overall health
of a portfolio after a sequence of trades, focusing on:
- Total Return (CAGR + dividends combined)
- Long-term Promise (consistency, financials, dividend stability)
- Stability (low volatility, minimal drawdown)

Used by the holistic planner to evaluate and compare action sequences.
"""

import logging
from typing import Dict, Tuple

# Import scorer from dedicated module
from app.modules.scoring.domain.scorers.end_state import score_total_return

logger = logging.getLogger(__name__)

# End-state scoring weights (default/balanced profile)
WEIGHT_TOTAL_RETURN = 0.35
WEIGHT_DIVERSIFICATION = 0.25
WEIGHT_LONG_TERM_PROMISE = 0.20
WEIGHT_STABILITY = 0.15
WEIGHT_OPINION = 0.05


def get_risk_profile_weights(risk_profile: str = "balanced") -> Dict[str, float]:
    """
    Get scoring weights adjusted for risk profile.

    Args:
        risk_profile: "conservative", "balanced", or "aggressive"

    Returns:
        Dict with weight_total_return, weight_diversification, weight_long_term_promise,
        weight_stability, weight_opinion (must sum to 1.0)
    """
    if risk_profile == "conservative":
        # Conservative: Emphasize stability and diversification, reduce return focus
        return {
            "weight_total_return": 0.25,
            "weight_diversification": 0.30,
            "weight_long_term_promise": 0.20,
            "weight_stability": 0.20,
            "weight_opinion": 0.05,
        }
    elif risk_profile == "aggressive":
        # Aggressive: Emphasize return and promise, reduce stability focus
        return {
            "weight_total_return": 0.45,
            "weight_diversification": 0.20,
            "weight_long_term_promise": 0.25,
            "weight_stability": 0.05,
            "weight_opinion": 0.05,
        }
    else:  # balanced (default)
        return {
            "weight_total_return": WEIGHT_TOTAL_RETURN,
            "weight_diversification": WEIGHT_DIVERSIFICATION,
            "weight_long_term_promise": WEIGHT_LONG_TERM_PROMISE,
            "weight_stability": WEIGHT_STABILITY,
            "weight_opinion": WEIGHT_OPINION,
        }


# Long-term promise sub-weights
PROMISE_WEIGHT_CONSISTENCY = 0.35
PROMISE_WEIGHT_FINANCIALS = 0.25
PROMISE_WEIGHT_DIVIDEND_STABILITY = 0.25
PROMISE_WEIGHT_SORTINO = 0.15

# Stability sub-weights
STABILITY_WEIGHT_VOLATILITY = 0.50
STABILITY_WEIGHT_DRAWDOWN = 0.30
STABILITY_WEIGHT_SHARPE = 0.20


async def calculate_total_return_score(
    symbol: str,
    metrics: Dict[str, float],
) -> Tuple[float, Dict[str, float]]:
    """
    Calculate total return score (CAGR + dividend yield combined).

    Args:
        symbol: Stock symbol (for reference only, not used for DB queries)
        metrics: Pre-fetched metrics dict containing CAGR_5Y and DIVIDEND_YIELD

    Returns:
        Tuple of (total_score, sub_components_dict)
    """
    # Get CAGR from metrics dict
    cagr = metrics.get("CAGR_5Y")
    if cagr is None:
        cagr = 0.0

    # Get dividend yield from metrics dict
    dividend_yield = metrics.get("DIVIDEND_YIELD")
    if dividend_yield is None:
        dividend_yield = 0.0

    # Calculate total return
    total_return = cagr + dividend_yield

    # Score it
    score = score_total_return(total_return)

    sub_components = {
        "cagr": round(cagr, 4),
        "dividend_yield": round(dividend_yield, 4),
        "total_return": round(total_return, 4),
        "score": round(score, 3),
    }

    return round(score, 3), sub_components


def _derive_dividend_consistency_from_payout(payout: float) -> float:
    """Derive dividend consistency score from payout ratio."""
    if 0.3 <= payout <= 0.6:
        return 1.0
    elif payout < 0.3:
        return 0.5 + (payout / 0.3) * 0.5
    elif payout <= 0.8:
        return 1.0 - ((payout - 0.6) / 0.2) * 0.3
    else:
        return 0.4


def _convert_sortino_to_score(sortino: float) -> float:
    """Convert Sortino ratio to score (0-1)."""
    if sortino >= 2.0:
        return 1.0
    elif sortino >= 1.5:
        return 0.8 + (sortino - 1.5) * 0.4
    elif sortino >= 1.0:
        return 0.6 + (sortino - 1.0) * 0.4
    elif sortino >= 0:
        return sortino * 0.6
    else:
        return 0.0


async def calculate_long_term_promise(
    symbol: str,
    metrics: Dict[str, float],
) -> Tuple[float, Dict[str, float]]:
    """
    Calculate long-term promise score.

    Combines:
    - Consistency (35%): 5Y vs 10Y CAGR similarity
    - Financial Strength (25%): Margins, debt, liquidity
    - Dividend Consistency (25%): No big cuts, sustainable payout
    - Sortino (15%): Good returns with low downside risk

    Args:
        symbol: Stock symbol (for reference only, not used for DB queries)
        metrics: Pre-fetched metrics dict containing CONSISTENCY_SCORE, FINANCIAL_STRENGTH,
                 DIVIDEND_CONSISTENCY (or PAYOUT_RATIO), and SORTINO

    Returns:
        Tuple of (total_score, sub_components_dict)
    """
    # Get consistency score from metrics
    consistency_score = metrics.get("CONSISTENCY_SCORE")
    if consistency_score is None:
        consistency_score = 0.5

    # Get financial strength from metrics
    financial_strength = metrics.get("FINANCIAL_STRENGTH")
    if financial_strength is None:
        financial_strength = 0.5

    # Get dividend consistency from metrics (or derive from payout ratio)
    dividend_consistency = metrics.get("DIVIDEND_CONSISTENCY")
    if dividend_consistency is None:
        payout = metrics.get("PAYOUT_RATIO")
        if payout is not None:
            dividend_consistency = _derive_dividend_consistency_from_payout(payout)
        else:
            dividend_consistency = 0.5

    # Get Sortino and convert to score
    sortino_raw = metrics.get("SORTINO")
    if sortino_raw is not None:
        sortino_score = _convert_sortino_to_score(sortino_raw)
    else:
        sortino_score = 0.5

    total = (
        consistency_score * PROMISE_WEIGHT_CONSISTENCY
        + financial_strength * PROMISE_WEIGHT_FINANCIALS
        + dividend_consistency * PROMISE_WEIGHT_DIVIDEND_STABILITY
        + sortino_score * PROMISE_WEIGHT_SORTINO
    )

    sub_components = {
        "consistency": round(consistency_score, 3),
        "financial_strength": round(financial_strength, 3),
        "dividend_consistency": round(dividend_consistency, 3),
        "sortino": round(sortino_score, 3),
    }

    return round(min(1.0, total), 3), sub_components


def _convert_volatility_to_score(volatility: float) -> float:
    """Convert volatility to score (inverse - lower is better)."""
    if volatility <= 0.15:
        return 1.0
    elif volatility <= 0.25:
        return 1.0 - ((volatility - 0.15) / 0.10) * 0.3
    elif volatility <= 0.40:
        return 0.7 - ((volatility - 0.25) / 0.15) * 0.4
    else:
        return max(0.1, 0.3 - (volatility - 0.40))


def _convert_drawdown_to_score(max_dd: float) -> float:
    """Convert max drawdown to score."""
    dd_pct = abs(max_dd)
    if dd_pct <= 0.10:
        return 1.0
    elif dd_pct <= 0.20:
        return 0.8 + (0.20 - dd_pct) * 2
    elif dd_pct <= 0.30:
        return 0.6 + (0.30 - dd_pct) * 2
    elif dd_pct <= 0.50:
        return 0.2 + (0.50 - dd_pct) * 2
    else:
        return max(0.0, 0.2 - (dd_pct - 0.50))


def _convert_sharpe_to_score(sharpe: float) -> float:
    """Convert Sharpe ratio to score."""
    if sharpe >= 2.0:
        return 1.0
    elif sharpe >= 1.0:
        return 0.7 + (sharpe - 1.0) * 0.3
    elif sharpe >= 0.5:
        return 0.4 + (sharpe - 0.5) * 0.6
    elif sharpe >= 0:
        return sharpe * 0.8
    else:
        return 0.0


async def calculate_stability_score(
    symbol: str,
    metrics: Dict[str, float],
) -> Tuple[float, Dict[str, float]]:
    """
    Calculate stability score.

    Combines:
    - Inverse Volatility (50%): Lower volatility = higher score
    - Drawdown Score (30%): Lower max drawdown = higher score
    - Sharpe Score (20%): Higher risk-adjusted returns = higher score

    Args:
        symbol: Stock symbol (for reference only, not used for DB queries)
        metrics: Pre-fetched metrics dict containing VOLATILITY_ANNUAL, MAX_DRAWDOWN, and SHARPE

    Returns:
        Tuple of (total_score, sub_components_dict)
    """
    # Get volatility and convert to score
    volatility_raw = metrics.get("VOLATILITY_ANNUAL")
    if volatility_raw is not None and volatility_raw > 0:
        volatility_score = _convert_volatility_to_score(volatility_raw)
    else:
        volatility_score = 0.5

    # Get drawdown and convert to score
    max_dd = metrics.get("MAX_DRAWDOWN")
    if max_dd is not None:
        drawdown_score = _convert_drawdown_to_score(max_dd)
    else:
        drawdown_score = 0.5

    # Get Sharpe and convert to score
    sharpe_raw = metrics.get("SHARPE")
    if sharpe_raw is not None:
        sharpe_score = _convert_sharpe_to_score(sharpe_raw)
    else:
        sharpe_score = 0.5

    total = (
        volatility_score * STABILITY_WEIGHT_VOLATILITY
        + drawdown_score * STABILITY_WEIGHT_DRAWDOWN
        + sharpe_score * STABILITY_WEIGHT_SHARPE
    )

    sub_components = {
        "volatility": round(volatility_score, 3),
        "drawdown": round(drawdown_score, 3),
        "sharpe": round(sharpe_score, 3),
    }

    return round(min(1.0, total), 3), sub_components


async def calculate_portfolio_end_state_score(
    positions: Dict[str, float],
    total_value: float,
    diversification_score: float,
    metrics_cache: Dict[str, Dict[str, float]],
    opinion_score: float = 0.5,
    risk_profile: str = "balanced",
) -> Tuple[float, Dict[str, float]]:
    """
    Calculate end-state score for an entire portfolio.

    This is the main function used by the holistic planner to evaluate
    the quality of a portfolio after executing a sequence of trades.

    Args:
        positions: Dict of symbol -> position value in EUR
        total_value: Total portfolio value in EUR
        diversification_score: Pre-calculated diversification score (0-1)
        metrics_cache: Pre-fetched metrics dict mapping symbol -> metrics dict
        opinion_score: Average analyst opinion score (default 0.5)

    Returns:
        Tuple of (total_score, detailed_breakdown)
    """
    if total_value <= 0 or not positions:
        return 0.5, {"error": "Invalid portfolio data"}  # type: ignore[dict-item]

    # Calculate weighted averages across all positions
    weighted_total_return = 0.0
    weighted_promise = 0.0
    weighted_stability = 0.0

    total_return_details = {}
    promise_details = {}
    stability_details = {}

    for symbol, value in positions.items():
        if value <= 0:
            continue

        weight = value / total_value

        # Get metrics for this symbol (use empty dict if missing)
        metrics = metrics_cache.get(symbol, {})

        # Get scores for this position using cached metrics
        tr_score, tr_subs = await calculate_total_return_score(symbol, metrics)
        promise_score, promise_subs = await calculate_long_term_promise(symbol, metrics)
        stab_score, stab_subs = await calculate_stability_score(symbol, metrics)

        weighted_total_return += tr_score * weight
        weighted_promise += promise_score * weight
        weighted_stability += stab_score * weight

        # Store details for transparency
        total_return_details[symbol] = {"score": tr_score, "weight": round(weight, 3)}
        promise_details[symbol] = {"score": promise_score, "weight": round(weight, 3)}
        stability_details[symbol] = {"score": stab_score, "weight": round(weight, 3)}

    # Get risk-adjusted weights
    weights = get_risk_profile_weights(risk_profile)
    weight_total_return = weights["weight_total_return"]
    weight_diversification = weights["weight_diversification"]
    weight_long_term_promise = weights["weight_long_term_promise"]
    weight_stability = weights["weight_stability"]
    weight_opinion = weights["weight_opinion"]

    # Calculate final end-state score with risk-adjusted weights
    end_state_score = (
        weighted_total_return * weight_total_return
        + diversification_score * weight_diversification
        + weighted_promise * weight_long_term_promise
        + weighted_stability * weight_stability
        + opinion_score * weight_opinion
    )

    detailed_breakdown = {
        "total_return": {
            "weighted_score": round(weighted_total_return, 3),
            "weight": weight_total_return,
            "contribution": round(weighted_total_return * weight_total_return, 3),
        },
        "diversification": {
            "score": round(diversification_score, 3),
            "weight": weight_diversification,
            "contribution": round(diversification_score * weight_diversification, 3),
        },
        "long_term_promise": {
            "weighted_score": round(weighted_promise, 3),
            "weight": weight_long_term_promise,
            "contribution": round(weighted_promise * weight_long_term_promise, 3),
        },
        "stability": {
            "weighted_score": round(weighted_stability, 3),
            "weight": weight_stability,
            "contribution": round(weighted_stability * weight_stability, 3),
        },
        "opinion": {
            "score": round(opinion_score, 3),
            "weight": weight_opinion,
            "contribution": round(opinion_score * weight_opinion, 3),
        },
        "end_state_score": round(end_state_score, 3),
        "risk_profile": risk_profile,
    }

    return round(min(1.0, end_state_score), 3), detailed_breakdown  # type: ignore[return-value]

File: trader-go/internal/modules/scoring/scorers/end_state.go
package scorers

import (
	"math"

	"github.com/aristath/arduino-trader/internal/modules/scoring"
)

// EndStateScorer provides scoring functions for holistic planning
// Faithful translation from Python: app/modules/scoring/domain/end_state.py
//
// This module provides scoring functions that evaluate the overall health
// of a portfolio after a sequence of trades, focusing on:
// - Total Return (CAGR + dividends combined)
// - Long-term Promise (consistency, financials, dividend stability)
// - Stability (low volatility, minimal drawdown)
//
// Used by the holistic planner to evaluate and compare action sequences.
type EndStateScorer struct{}

// EndStateWeights represents risk-adjusted weights for end-state scoring
type EndStateWeights struct {
	WeightTotalReturn     float64 `json:"weight_total_return"`
	WeightDiversification float64 `json:"weight_diversification"`
	WeightLongTermPromise float64 `json:"weight_long_term_promise"`
	WeightStability       float64 `json:"weight_stability"`
	WeightOpinion         float64 `json:"weight_opinion"`
}

// TotalReturnScore represents total return scoring components
type TotalReturnScore struct {
	CAGR          float64 `json:"cagr"`
	DividendYield float64 `json:"dividend_yield"`
	TotalReturn   float64 `json:"total_return"`
	Score         float64 `json:"score"`
}

// LongTermPromiseScore represents long-term promise components
type LongTermPromiseScore struct {
	Consistency         float64 `json:"consistency"`
	FinancialStrength   float64 `json:"financial_strength"`
	DividendConsistency float64 `json:"dividend_consistency"`
	Sortino             float64 `json:"sortino"`
	Score               float64 `json:"score"`
}

// StabilityScore represents stability components
type StabilityScore struct {
	Volatility float64 `json:"volatility"`
	Drawdown   float64 `json:"drawdown"`
	Sharpe     float64 `json:"sharpe"`
	Score      float64 `json:"score"`
}

// PortfolioEndStateScore represents complete end-state analysis
type PortfolioEndStateScore struct {
	RiskProfile string `json:"risk_profile"`
	Error       string `json:"error,omitempty"`
	TotalReturn struct {
		WeightedScore float64 `json:"weighted_score"`
		Weight        float64 `json:"weight"`
		Contribution  float64 `json:"contribution"`
	} `json:"total_return"`
	Diversification struct {
		Score        float64 `json:"score"`
		Weight       float64 `json:"weight"`
		Contribution float64 `json:"contribution"`
	} `json:"diversification"`
	LongTermPromise struct {
		WeightedScore float64 `json:"weighted_score"`
		Weight        float64 `json:"weight"`
		Contribution  float64 `json:"contribution"`
	} `json:"long_term_promise"`
	Stability struct {
		WeightedScore float64 `json:"weighted_score"`
		Weight        float64 `json:"weight"`
		Contribution  float64 `json:"contribution"`
	} `json:"stability"`
	Opinion struct {
		Score        float64 `json:"score"`
		Weight       float64 `json:"weight"`
		Contribution float64 `json:"contribution"`
	} `json:"opinion"`
	EndStateScore float64 `json:"end_state_score"`
}

// End-state scoring weights (default/balanced profile)
const (
	weightTotalReturnDefault     = 0.35
	weightDiversificationDefault = 0.25
	weightLongTermPromiseDefault = 0.20
	weightStabilityDefault       = 0.15
	weightOpinionDefault         = 0.05

	// Long-term promise sub-weights
	promiseWeightConsistency       = 0.35
	promiseWeightFinancials        = 0.25
	promiseWeightDividendStability = 0.25
	promiseWeightSortino           = 0.15

	// Stability sub-weights
	stabilityWeightVolatility = 0.50
	stabilityWeightDrawdown   = 0.30
	stabilityWeightSharpe     = 0.20
)

// NewEndStateScorer creates a new end-state scorer
func NewEndStateScorer() *EndStateScorer {
	return &EndStateScorer{}
}

// GetRiskProfileWeights returns scoring weights adjusted for risk profile
//
// Args:
//
//	riskProfile: "conservative", "balanced", or "aggressive"
//
// Returns:
//
//	EndStateWeights with weights that sum to 1.0
func (es *EndStateScorer) GetRiskProfileWeights(riskProfile string) EndStateWeights {
	switch riskProfile {
	case "conservative":
		// Conservative: Emphasize stability and diversification, reduce return focus
		return EndStateWeights{
			WeightTotalReturn:     0.25,
			WeightDiversification: 0.30,
			WeightLongTermPromise: 0.20,
			WeightStability:       0.20,
			WeightOpinion:         0.05,
		}
	case "aggressive":
		// Aggressive: Emphasize return and promise, reduce stability focus
		return EndStateWeights{
			WeightTotalReturn:     0.45,
			WeightDiversification: 0.20,
			WeightLongTermPromise: 0.25,
			WeightStability:       0.05,
			WeightOpinion:         0.05,
		}
	default: // balanced
		return EndStateWeights{
			WeightTotalReturn:     weightTotalReturnDefault,
			WeightDiversification: weightDiversificationDefault,
			WeightLongTermPromise: weightLongTermPromiseDefault,
			WeightStability:       weightStabilityDefault,
			WeightOpinion:         weightOpinionDefault,
		}
	}
}

// ScoreTotalReturn calculates bell curve scoring for total return
//
// Peak at target (default 12%). Uses asymmetric Gaussian.
//
// Args:
//
//	totalReturn: Total return value
//	target: Target return (default 12%)
//
// Returns:
//
//	Score from 0.15 to 1.0
func (es *EndStateScorer) ScoreTotalReturn(totalReturn, target float64) float64 {
	if totalReturn <= 0 {
		return scoring.BellCurveFloor
	}

	// Use slightly wider sigma for total return (more forgiving than CAGR)
	var sigma float64
	if totalReturn < target {
		sigma = scoring.BellCurveSigmaLeft
	} else {
		sigma = scoring.BellCurveSigmaRight * 1.2
	}

	rawScore := math.Exp(-math.Pow(totalReturn-target, 2) / (2 * math.Pow(sigma, 2)))

	return scoring.BellCurveFloor + rawScore*(1-scoring.BellCurveFloor)
}

// CalculateTotalReturnScore calculates total return score (CAGR + dividend yield combined)
//
// Args:
//
//	metrics: Pre-fetched metrics dict containing CAGR_5Y and DIVIDEND_YIELD
//
// Returns:
//
//	TotalReturnScore with all components
func (es *EndStateScorer) CalculateTotalReturnScore(metrics map[string]float64) TotalReturnScore {
	// Get CAGR from metrics dict
	cagr, hasCagr := metrics["CAGR_5Y"]
	if !hasCagr {
		cagr = 0.0
	}

	// Get dividend yield from metrics dict
	dividendYield, hasDiv := metrics["DIVIDEND_YIELD"]
	if !hasDiv {
		dividendYield = 0.0
	}

	// Calculate total return
	totalReturn := cagr + dividendYield

	// Score it
	score := es.ScoreTotalReturn(totalReturn, 0.12)

	return TotalReturnScore{
		CAGR:          round4(cagr),
		DividendYield: round4(dividendYield),
		TotalReturn:   round4(totalReturn),
		Score:         round3(score),
	}
}

// deriveDividendConsistencyFromPayout derives dividend consistency score from payout ratio
func deriveDividendConsistencyFromPayout(payout float64) float64 {
	if payout >= 0.3 && payout <= 0.6 {
		return 1.0
	} else if payout < 0.3 {
		return 0.5 + (payout/0.3)*0.5
	} else if payout <= 0.8 {
		return 1.0 - ((payout-0.6)/0.2)*0.3
	}
	return 0.4
}

// convertSortinoToScore converts Sortino ratio to score (0-1)
func convertSortinoToScore(sortino float64) float64 {
	if sortino >= 2.0 {
		return 1.0
	} else if sortino >= 1.5 {
		return 0.8 + (sortino-1.5)*0.4
	} else if sortino >= 1.0 {
		return 0.6 + (sortino-1.0)*0.4
	} else if sortino >= 0 {
		return sortino * 0.6
	}
	return 0.0
}

// CalculateLongTermPromise calculates long-term promise score
//
// Combines:
// - Consistency (35%): 5Y vs 10Y CAGR similarity
// - Financial Strength (25%): Margins, debt, liquidity
// - Dividend Consistency (25%): No big cuts, sustainable payout
// - Sortino (15%): Good returns with low downside risk
//
// Args:
//
//	metrics: Pre-fetched metrics dict containing CONSISTENCY_SCORE, FINANCIAL_STRENGTH,
//	         DIVIDEND_CONSISTENCY (or PAYOUT_RATIO), and SORTINO
//
// Returns:
//
//	LongTermPromiseScore with all components
func (es *EndStateScorer) CalculateLongTermPromise(metrics map[string]float64) LongTermPromiseScore {
	// Get consistency score from metrics
	consistencyScore, hasConsistency := metrics["CONSISTENCY_SCORE"]
	if !hasConsistency {
		consistencyScore = 0.5
	}

	// Get financial strength from metrics
	financialStrength, hasFinancial := metrics["FINANCIAL_STRENGTH"]
	if !hasFinancial {
		financialStrength = 0.5
	}

	// Get dividend consistency from metrics (or derive from payout ratio)
	dividendConsistency, hasDivConsistency := metrics["DIVIDEND_CONSISTENCY"]
	if !hasDivConsistency {
		payout, hasPayout := metrics["PAYOUT_RATIO"]
		if hasPayout {
			dividendConsistency = deriveDividendConsistencyFromPayout(payout)
		} else {
			dividendConsistency = 0.5
		}
	}

	// Get Sortino and convert to score
	sortinoScore := 0.5
	sortinoRaw, hasSortino := metrics["SORTINO"]
	if hasSortino {
		sortinoScore = convertSortinoToScore(sortinoRaw)
	}

	total := consistencyScore*promiseWeightConsistency +
		financialStrength*promiseWeightFinancials +
		dividendConsistency*promiseWeightDividendStability +
		sortinoScore*promiseWeightSortino

	return LongTermPromiseScore{
		Consistency:         round3(consistencyScore),
		FinancialStrength:   round3(financialStrength),
		DividendConsistency: round3(dividendConsistency),
		Sortino:             round3(sortinoScore),
		Score:               round3(math.Min(1.0, total)),
	}
}

// convertVolatilityToScore converts volatility to score (inverse - lower is better)
func convertVolatilityToScore(volatility float64) float64 {
	if volatility <= 0.15 {
		return 1.0
	} else if volatility <= 0.25 {
		return 1.0 - ((volatility-0.15)/0.10)*0.3
	} else if volatility <= 0.40 {
		return 0.7 - ((volatility-0.25)/0.15)*0.4
	}
	return math.Max(0.1, 0.3-(volatility-0.40))
}

// convertDrawdownToScore converts max drawdown to score
func convertDrawdownToScore(maxDD float64) float64 {
	ddPct := math.Abs(maxDD)
	if ddPct <= 0.10 {
		return 1.0
	} else if ddPct <= 0.20 {
		return 0.8 + (0.20-ddPct)*2
	} else if ddPct <= 0.30 {
		return 0.6 + (0.30-ddPct)*2
	} else if ddPct <= 0.50 {
		return 0.2 + (0.50-ddPct)*2
	}
	return math.Max(0.0, 0.2-(ddPct-0.50))
}

// convertSharpeToScore converts Sharpe ratio to score
func convertSharpeToScore(sharpe float64) float64 {
	if sharpe >= 2.0 {
		return 1.0
	} else if sharpe >= 1.0 {
		return 0.7 + (sharpe-1.0)*0.3
	} else if sharpe >= 0.5 {
		return 0.4 + (sharpe-0.5)*0.6
	} else if sharpe >= 0 {
		return sharpe * 0.8
	}
	return 0.0
}

// CalculateStabilityScore calculates stability score
//
// Combines:
// - Inverse Volatility (50%): Lower volatility = higher score
// - Drawdown Score (30%): Lower max drawdown = higher score
// - Sharpe Score (20%): Higher risk-adjusted returns = higher score
//
// Args:
//
//	metrics: Pre-fetched metrics dict containing VOLATILITY_ANNUAL, MAX_DRAWDOWN, and SHARPE
//
// Returns:
//
//	StabilityScore with all components
func (es *EndStateScorer) CalculateStabilityScore(metrics map[string]float64) StabilityScore {
	// Get volatility and convert to score
	volatilityScore := 0.5
	volatilityRaw, hasVol := metrics["VOLATILITY_ANNUAL"]
	if hasVol && volatilityRaw > 0 {
		volatilityScore = convertVolatilityToScore(volatilityRaw)
	}

	// Get drawdown and convert to score
	drawdownScore := 0.5
	maxDD, hasDD := metrics["MAX_DRAWDOWN"]
	if hasDD {
		drawdownScore = convertDrawdownToScore(maxDD)
	}

	// Get Sharpe and convert to score
	sharpeScore := 0.5
	sharpeRaw, hasSharpe := metrics["SHARPE"]
	if hasSharpe {
		sharpeScore = convertSharpeToScore(sharpeRaw)
	}

	total := volatilityScore*stabilityWeightVolatility +
		drawdownScore*stabilityWeightDrawdown +
		sharpeScore*stabilityWeightSharpe

	return StabilityScore{
		Volatility: round3(volatilityScore),
		Drawdown:   round3(drawdownScore),
		Sharpe:     round3(sharpeScore),
		Score:      round3(math.Min(1.0, total)),
	}
}

// CalculatePortfolioEndStateScore calculates end-state score for an entire portfolio
//
// This is the main function used by the holistic planner to evaluate
// the quality of a portfolio after executing a sequence of trades.
//
// Args:
//
//	positions: Map of symbol -> position value in EUR
//	totalValue: Total portfolio value in EUR
//	diversificationScore: Pre-calculated diversification score (0-1)
//	metricsCache: Pre-fetched metrics dict mapping symbol -> metrics dict
//	opinionScore: Average analyst opinion score (default 0.5)
//	riskProfile: "conservative", "balanced", or "aggressive"
//
// Returns:
//
//	PortfolioEndStateScore with detailed breakdown
func (es *EndStateScorer) CalculatePortfolioEndStateScore(
	positions map[string]float64,
	totalValue float64,
	diversificationScore float64,
	metricsCache map[string]map[string]float64,
	opinionScore float64,
	riskProfile string,
) PortfolioEndStateScore {
	result := PortfolioEndStateScore{
		RiskProfile: riskProfile,
	}

	if totalValue <= 0 || len(positions) == 0 {
		result.Error = "Invalid portfolio data"
		result.EndStateScore = 0.5
		return result
	}

	// Calculate weighted averages across all positions
	weightedTotalReturn := 0.0
	weightedPromise := 0.0
	weightedStability := 0.0

	for symbol, value := range positions {
		if value <= 0 {
			continue
		}

		weight := value / totalValue

		// Get metrics for this symbol (use empty dict if missing)
		metrics, hasMetrics := metricsCache[symbol]
		if !hasMetrics {
			metrics = make(map[string]float64)
		}

		// Get scores for this position using cached metrics
		trScore := es.CalculateTotalReturnScore(metrics)
		promiseScore := es.CalculateLongTermPromise(metrics)
		stabScore := es.CalculateStabilityScore(metrics)

		weightedTotalReturn += trScore.Score * weight
		weightedPromise += promiseScore.Score * weight
		weightedStability += stabScore.Score * weight
	}

	// Get risk-adjusted weights
	weights := es.GetRiskProfileWeights(riskProfile)

	// Calculate final end-state score with risk-adjusted weights
	endStateScore := weightedTotalReturn*weights.WeightTotalReturn +
		diversificationScore*weights.WeightDiversification +
		weightedPromise*weights.WeightLongTermPromise +
		weightedStability*weights.WeightStability +
		opinionScore*weights.WeightOpinion

	// Build detailed breakdown
	result.TotalReturn.WeightedScore = round3(weightedTotalReturn)
	result.TotalReturn.Weight = weights.WeightTotalReturn
	result.TotalReturn.Contribution = round3(weightedTotalReturn * weights.WeightTotalReturn)

	result.Diversification.Score = round3(diversificationScore)
	result.Diversification.Weight = weights.WeightDiversification
	result.Diversification.Contribution = round3(diversificationScore * weights.WeightDiversification)

	result.LongTermPromise.WeightedScore = round3(weightedPromise)
	result.LongTermPromise.Weight = weights.WeightLongTermPromise
	result.LongTermPromise.Contribution = round3(weightedPromise * weights.WeightLongTermPromise)

	result.Stability.WeightedScore = round3(weightedStability)
	result.Stability.Weight = weights.WeightStability
	result.Stability.Contribution = round3(weightedStability * weights.WeightStability)

	result.Opinion.Score = round3(opinionScore)
	result.Opinion.Weight = weights.WeightOpinion
	result.Opinion.Contribution = round3(opinionScore * weights.WeightOpinion)

	result.EndStateScore = round3(math.Min(1.0, endStateScore))

	return result
}
